### PR TITLE
fix: preventing original chain name & destination chain name are the same

### PR DESCRIPTION
**Pull request by ahramy**
[AXE-4776](https://axelarnetwork.atlassian.net/browse/AXE-4776)

- In `InterchainTokenFactory.deployRemoteInterchainToken`, check if `destinationChain` is the same as `originalChainName` to prevent remote deployment requests to the original chain.
- Check originalChainName and destinationChain in `InterchainTokenFactory.deployRemoteCanonicalInterchainToken` to ensure that a remote deploy cannot be requested to the original chain.
- Unit test coverage is 100% for `InterchainTokenFactory.sol` and `InterchainTokenService.sol`
```
----------------------------------------|----------|----------|----------|----------|----------------|
File                                    |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
----------------------------------------|----------|----------|----------|----------|----------------|
  InterchainTokenFactory.sol            |      100 |      100 |      100 |      100 |                |
  InterchainTokenService.sol            |      100 |      100 |      100 |      100 |                |

```

[AXE-4776]: https://axelarnetwork.atlassian.net/browse/AXE-4776?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: contracts/InterchainTokenFactory.sol [hunk 1]
```diff
@@ -285,6 +286,8 @@ contract InterchainTokenFactory is IInterchainTokenFactory, ITokenManagerType, M
             } else {
                 chainNameHash_ = keccak256(bytes(originalChain));
             }
+            if (chainNameHash_ == keccak256(bytes(destinationChain))) revert CannotDeployRemotely(destinationChain);
```
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:289]: I think we want to remove the ability to deploy a token with a different `originalChain` instead. And perform this check inside ITS `deploy*` methods instead so you can never submit a self deploy call

**Review by Foivos — APPROVED** [review]

File: contracts/InterchainTokenFactory.sol (hunk 1 above)
Comment (ahramy) [reply contracts/InterchainTokenFactory.sol:289]: Had offline discussion. updated.

File: contracts/InterchainTokenFactory.sol [hunk 2]
```diff
@@ -191,7 +191,7 @@ contract InterchainTokenFactory is IInterchainTokenFactory, ITokenManagerType, M
             if (bytes(originalChainName).length == 0) {
                 chainNameHash_ = chainNameHash;
             } else {
-                chainNameHash_ = keccak256(bytes(originalChainName));
+                revert NotSupported();
```
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:194]: behaviour should be documented

File: contracts/InterchainTokenFactory.sol [hunk 3]
```diff
@@ -283,8 +283,9 @@ contract InterchainTokenFactory is IInterchainTokenFactory, ITokenManagerType, M
             if (bytes(originalChain).length == 0) {
                 chainNameHash_ = chainNameHash;
             } else {
-                chainNameHash_ = keccak256(bytes(originalChain));
+                revert NotSupported();
```
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:286]: same

File: contracts/interfaces/IInterchainTokenService.sol [hunk 4]
```diff
@@ -48,6 +48,7 @@ interface IInterchainTokenService is
     error PostDeployFailed(bytes data);
     error ZeroAmount();
     error CannotDeploy(TokenManagerType);
+    error CannotDeployRemotely(string chainName);
```
Comment (milapsheth) [inline contracts/interfaces/IInterchainTokenService.sol:51]:
```suggestion
    error CannotDeployRemotelyToSelf();
```

File: contracts/InterchainTokenFactory.sol (hunk 2 above)
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:194]: add unit test coverage

File: contracts/InterchainTokenFactory.sol (hunk 3 above)
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:286]: an overloaded version of these methods should be created without the `originalChain` arg

File: contracts/InterchainTokenFactory.sol [hunk 5]
```diff
@@ -215,6 +215,49 @@ contract InterchainTokenFactory is IInterchainTokenFactory, ITokenManagerType, M
         tokenId = _deployInterchainToken(salt, destinationChain, tokenName, tokenSymbol, tokenDecimals, minter_, gasValue);
     }
 
+    /**
+     * @notice The deployRemoteInterchainToken with originalChainName allows deploying a remote interchain token back to the same chain, which could cause issues with ITS Hub balance tracking.
+     * To fix this, the deployRemoteInterchainToken function was overloaded to prevent self-deployemnt while still allowing self-transfers.
```
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:220]:
```suggestion
     * @notice Deploys a remote interchain token on a specified destination chain.
```
Should document what the method does. This method doesn't need to have any discussions about originalChain. It should be discussed in the other `deployRemoteInterchainToken` (and as a `@dev` comment). Put this method first, so users will see it first and use this by default.

This part of the comment `... which could cause issues with ITS Hub balance tracking ...` is vague and so doesn't make much sense to external users. It's just an internal note, it would be better to mention the larger point, something like: `originalChainName is only allowed to be '', i.e the current chain. Other source chains are not supported anymore to simplify ITS deployment behaviour.`

File: contracts/InterchainTokenFactory.sol [hunk 6]
```diff
@@ -215,6 +215,49 @@ contract InterchainTokenFactory is IInterchainTokenFactory, ITokenManagerType, M
...
+        uint256 gasValue
+    ) external payable returns (bytes32 tokenId) {
+        string memory tokenName;
+        string memory tokenSymbol;
+        uint8 tokenDecimals;
+        bytes memory minter_ = new bytes(0);
+
+        {
+            address sender = msg.sender;
+            salt = interchainTokenSalt(chainNameHash, sender, salt);
```
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:241]: chainNameHash is not defined

File: contracts/InterchainTokenFactory.sol [hunk 7]
```diff
@@ -215,6 +215,49 @@ contract InterchainTokenFactory is IInterchainTokenFactory, ITokenManagerType, M
...
+            tokenId = interchainTokenService.interchainTokenId(TOKEN_FACTORY_DEPLOYER, salt);
```
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:242]: avoid duplication. Call this method from the other deployRemoteInterchainToken after performing the originalChainName check

File: contracts/InterchainTokenFactory.sol [hunk 8]
```diff
@@ -299,6 +343,37 @@ contract InterchainTokenFactory is IInterchainTokenFactory, ITokenManagerType, M
...
 
+    /**
+     * @notice The deployRemoteCanonicalInterchainToken with originalChain allows deploying a canonical interchain token back to the same chain, which could cause issues with ITS Hub balance tracking.
+     * To fix this, the deployRemoteCanonicalInterchainToken function was overloaded to prevent self-deployemnt while still allowing self-transfers.
+     * @param originalTokenAddress The address of the original token on the original chain.
+     * @param destinationChain The name of the chain where the token will be deployed.
+     * @param gasValue The gas amount to be sent for deployment.
+     * @return tokenId The tokenId corresponding to the deployed InterchainToken.
+     */
+    function deployRemoteCanonicalInterchainToken(
```
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:354]: same comments as above

File: contracts/interfaces/IInterchainTokenFactory.sol [hunk 9]
```diff
@@ -131,6 +147,20 @@ interface IInterchainTokenFactory is IUpgradable, IMulticall {
         uint256 gasValue
     ) external payable returns (bytes32 tokenId);
 
+    /**
+     * @notice The deployRemoteCanonicalInterchainToken with originalChain allows deploying a canonical interchain token back to the same chain, which could cause issues with ITS Hub balance tracking.
```
Comment (milapsheth) [inline contracts/interfaces/IInterchainTokenFactory.sol:151]: update docstring and order as per other comments

**Review by milapsheth — CHANGES_REQUESTED** [review]

File: contracts/InterchainTokenFactory.sol (hunk 5 above)
Comment (ahramy) [reply contracts/InterchainTokenFactory.sol]: Thanks for comment. updated.

File: contracts/InterchainTokenFactory.sol [hunk 10]
```diff
@@ -175,27 +174,19 @@ contract InterchainTokenFactory is IInterchainTokenFactory, ITokenManagerType, M
...
         string memory destinationChain,
         uint256 gasValue
-    ) external payable returns (bytes32 tokenId) {
+    ) public payable returns (bytes32 tokenId) {
         string memory tokenName;
         string memory tokenSymbol;
         uint8 tokenDecimals;
         bytes memory minter_ = new bytes(0);
 
         {
```
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:187]: wondering if this additional scoping is still needed now that we removed one variable (maybe stack too deep error goes away)

File: contracts/InterchainTokenFactory.sol [hunk 11]
```diff
@@ -216,8 +207,10 @@ contract InterchainTokenFactory is IInterchainTokenFactory, ITokenManagerType, M
     }
 
     /**
-     * @notice The deployRemoteInterchainToken with originalChainName allows deploying a remote interchain token back to the same chain, which could cause issues with ITS Hub balance tracking.
-     * To fix this, the deployRemoteInterchainToken function was overloaded to prevent self-deployemnt while still allowing self-transfers.
+     * @notice Deploys a remote interchain token on a specified destination chain.
```
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:210]: Is there a `@deprecated` notion in natspec?
```suggestion
     * @notice Deploys a remote interchain token on a specified destination chain. This method is deprecated and will be removed in the future. Please use the above method instead.
```

File: contracts/InterchainTokenFactory.sol [hunk 12]
```diff
@@ -216,8 +207,10 @@ contract InterchainTokenFactory is IInterchainTokenFactory, ITokenManagerType, M
...
+     * @dev originalChainName is only allowed to be '', i.e the current chain.
+     * Other source chains are not supported anymore to simplify ITS deployment behaviour.
```
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:212]:
```suggestion
     * Other source chains are not supported anymore to simplify ITS token deployment behaviour.
```

File: contracts/InterchainTokenFactory.sol [hunk 13]
```diff
@@ -226,36 +219,14 @@ contract InterchainTokenFactory is IInterchainTokenFactory, ITokenManagerType, M
...
-                if (!token.isMinter(minter)) revert NotMinter(minter);
-                if (minter == address(interchainTokenService)) revert InvalidMinter(minter);
-
-                minter_ = minter.toBytes();
-            }
-        }
-
-        tokenId = _deployInterchainToken(salt, destinationChain, tokenName, tokenSymbol, tokenDecimals, minter_, gasValue);
+        if (bytes(originalChainName).length != 0) revert NotSupported();
+        tokenId = deployRemoteInterchainToken(salt, minter, destinationChain, gasValue);
```
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:229]:
```suggestion
        if (bytes(originalChainName).length != 0) revert NotSupported();

        tokenId = deployRemoteInterchainToken(salt, minter, destinationChain, gasValue);
```
separating checks from the logic

File: contracts/InterchainTokenFactory.sol [hunk 14]
```diff
@@ -306,31 +277,22 @@ contract InterchainTokenFactory is IInterchainTokenFactory, ITokenManagerType, M
...
-        string calldata originalChain,
         address originalTokenAddress,
         string calldata destinationChain,
         uint256 gasValue
-    ) external payable returns (bytes32 tokenId) {
+    ) public payable returns (bytes32 tokenId) {
         bytes32 salt;
         IInterchainToken token;
 
         {
```
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:293]: same comment

File: contracts/InterchainTokenFactory.sol [hunk 15]
```diff
@@ -344,34 +306,23 @@ contract InterchainTokenFactory is IInterchainTokenFactory, ITokenManagerType, M
     }
 
     /**
-     * @notice The deployRemoteCanonicalInterchainToken with originalChain allows deploying a canonical interchain token back to the same chain, which could cause issues with ITS Hub balance tracking.
-     * To fix this, the deployRemoteCanonicalInterchainToken function was overloaded to prevent self-deployemnt while still allowing self-transfers.
+     * @notice Deploys a canonical interchain token on a remote chain.
```
Comment (milapsheth) [inline contracts/InterchainTokenFactory.sol:309]: same comments

File: test/InterchainTokenFactory.js [hunk 16]
```diff
@@ -449,14 +446,31 @@ describe('InterchainTokenFactory', () => {
                         },
                     ),
                 tokenFactory,
+                'NotSupported',
+            );
+
+            await expectRevert(
+                (gasOptions) =>
+                    tokenFactory['deployRemoteInterchainToken(string,bytes32,address,string,uint256)'](
+                        '',
```
Comment (milapsheth) [inline test/InterchainTokenFactory.js:455]: add a test to ensure it fails if original Chain is not empty (double check test coverage for the repo, so it doesn't drop when making a PR)

File: contracts/InterchainTokenFactory.sol (hunk 10 above)
Comment (ahramy) [reply contracts/InterchainTokenFactory.sol]: That's a great call, actually it works without additional scoping. Updated.

File: contracts/InterchainTokenFactory.sol (hunk 11 above)
Comment (ahramy) [reply contracts/InterchainTokenFactory.sol:207]: There's no `@deprecated` notion in Natspec: https://docs.soliditylang.org/en/latest/natspec-format.html#tags
Updated it as the suggestion.

File: contracts/InterchainTokenFactory.sol (hunk 12 above)
Comment (ahramy) [reply contracts/InterchainTokenFactory.sol]: updated.

File: contracts/InterchainTokenFactory.sol (hunk 13 above)
Comment (ahramy) [reply contracts/InterchainTokenFactory.sol]: updated.

File: contracts/InterchainTokenFactory.sol (hunk 14 above)
Comment (ahramy) [reply contracts/InterchainTokenFactory.sol]: updated.

File: contracts/InterchainTokenFactory.sol (hunk 15 above)
Comment (ahramy) [reply contracts/InterchainTokenFactory.sol:306]: updated.

File: test/InterchainTokenFactory.js (hunk 16 above)
Comment (ahramy) [reply test/InterchainTokenFactory.js:455]: `Fails if original chain is not empty` is tested in line 449 with `chainName == "Test"`.
```
await expectRevert(
                (gasOptions) =>
                    tokenFactory['deployRemoteInterchainToken(string,bytes32,address,string,uint256)'](
                        chainName,
                        salt,
                        otherWallet.address,
                        destinationChain,
                        gasValue,
                        {
                            ...gasOptions,
                            value: gasValue,
                        },
                    ),
                tokenFactory,
                'NotSupported',
            );
```
Test coverage for `InterchainTokenFactory.sol` and `InterchainTokenService.sol ` are all 100%. Is there something I'm missing about the test case?

**Review by milapsheth — APPROVED** [review]
